src/Generator/Passes/DelegatesPass.cs

```diff
@@ -35,33 +42,135 @@ public DelegatesPass()
+            var typedef = FindMatchingTypedef(typedefs, functionType);
+            if (typedef == null)
+            {
+                for (int i = 0; i < functionType.Parameters.Count; i++)
+                {
+                    functionType.Parameters[i].Name = string.Format("_{0}", i);
+                }
+
+                if (Options.IsCSharpGenerator)
+                    return AddDelegatesToDictionary(@namespace, decl);
+
+                var delegateName = GenerateDelegateSignature(functionType.Parameters, functionType.ReturnType);
+
+                typedef = new TypedefDecl
+                {
+                    Access = AccessSpecifier.Public,
+                    Name = delegateName,
+                    Namespace = @namespace,
+                    QualifiedType = type,
+                    IsSynthetized = true
+                };
+                typedefs.Add(new Typedef
+                {
+                    Context = @namespace,
+                    Declaration = typedef
+                });
+            }
+
+            var typedefType = new TypedefType
+            {
+                Declaration = typedef
+            };
+            return new QualifiedType(typedefType);
+        }
+
         public override bool VisitASTContext(ASTContext context)
         {
-            foreach (var module in Options.Modules)
-                libsDelegates[module] = new Dictionary<string, DelegateDefinition>();
+            bool result = base.VisitASTContext(context);
+
+            foreach (var typedef in allTypedefs)
+            {
+                foreach (var foo in typedef.Value)
+                {
+                    foo.Context.Declarations.Add(foo.Declaration);
+                }
+            }
+            allTypedefs.Clear();
 
-            var unit = context.TranslationUnits.GetGenerated().LastOrDefault();
+            if (Options.IsCSharpGenerator)
+            {
+                foreach (var module in Options.Modules)
+                    libsDelegates[module] = new Dictionary<string, DelegateDefinition>();
 
-            if (unit == null)
-                return false;
+                var unit = context.TranslationUnits.GetGenerated().LastOrDefault();
 
-            var result = base.VisitASTContext(context);
+                if (unit == null)
+                    return false;
 
-            foreach (var module in Options.Modules.Where(m => namespacesDelegates.ContainsKey(m)))
-                namespacesDelegates[module].Namespace.Declarations.Add(namespacesDelegates[module]);
+                foreach (var module in Options.Modules.Where(m => namespacesDelegates.ContainsKey(m)))
+                    namespacesDelegates[module].Namespace.Declarations.Add(namespacesDelegates[module]);
+            }
 
             return result;
         }
 
-        private void AddDelegatesToDictionary(Declaration decl, Module module)
+        private QualifiedType AddDelegatesToDictionary(DeclarationContext @namespace, Declaration decl)
```

Comment:
This method adds at most one delegate so why the plural? Wouldn't AddDelegateToDictionary make more sense?